connect/api/src/main/java/org/apache/kafka/connect/data/Values.java

```diff
@@ -766,135 +852,23 @@ protected static boolean canParseSingleTokenLiteral(Parser parser, boolean embed
     protected static SchemaAndValue parse(Parser parser, boolean embedded) throws NoSuchElementException {
```

Comment:
That's a cool idea, I think that makes a lot of sense when a bunch of these static methods take a Parser argument anyway.

Since this is in the public API, I'll focus on moving some of the internal methods to instance methods of a protected/package local Parser, and leave the public static methods in Values.

Reply:
Ah right, Parser is also part of the public API, I thought it was only a private inner class.

Reply:
Parser was protected, so I think it's still safe to refactor. The class doesn't show up here: https://javadoc.io/doc/org.apache.kafka/connect-api/latest/index.html

I moved the existing Parser to Tokenizer, as it had a good interface already, and adding methods would just be clutter. The methods which took a Parser argument have now been moved to a single toplevel class named Parser. Both of these classes are package-local, so shouldn't appear in the API docs.

I left almost all of the private/protected static methods in Values, just bringing a few over that were only ever called by the Parser. I tried moving things from Values to Parser to break the circular dependency, but this required moving nearly everything to Parser. The two classes are really intertwined, and i'm not really satisfied with this refactor now.

Reply:
My core difficulty is that the parsing logic and the conversion logic mutually depend on one another: 
1. The convertTo methods check if the input is a String, and then run it through the Parser.
2. After parsing a map or array, the Parser calls convertTo on the elements to "cast" them to a common schema

I'm pretty sure convertTo -> parser -> convertTo is a reasonable cycle, and should happen all the time via convertToList, convertToMap.

I don't think that parser -> convertTo -> parser is a useful cycle for multiple reasons, but proving that is a little bit slippery. With some time I think I can break this part of the cycle so that this doesn't end up as one big ball of code again.

Reply:
I decided that pulling this class apart into standalone classes is possible, but a little bit risky compatibility-wise, and more complexity than I wanted to take on in this PR. I've moved the static methods accepting Parser to be instance methods of an inner class, so that they can still call into Values to do conversions.

Reply:
Regarding the parser -> convertTo -> parser cycle, if we want to tackle it, it's definitively in another PR.

Reply:
I'm confused. Locally for me this doesn't appear in the API docs, and it's because of this change: https://issues.apache.org/jira/browse/KAFKA-14839 . The ticket says it's fixed in 3.5.0, and that commit is definitely in 3.6.0+ so it should have applied when doing the 3.7.0 release.

I'm happy to preserve all of the protected methods (both the existing ones in Values, and the whole Parser class) if these are truly public API, but if this other commit already removed them from the Javadoc, I think we should try and follow through on the renaming here. I just need to figure out how they showed up in the 37 release docs...

Reply:
Oh it's because there are two javadoc configs, and they got out-of-sync:
`./gradlew javadoc` uses this one: https://github.com/apache/kafka/blob/5439914c32fa00d634efa7219699f1bc21add839/build.gradle#L597-L610
and `./greadlew aggregatedJavadoc` uses this one:
https://github.com/apache/kafka/blob/5439914c32fa00d634efa7219699f1bc21add839/build.gradle#L3335-L3353

Reply:
Good find! In that case I'm fine with the changes, no need for a KIP.